### ISSUE TITLE
load audio with torchaudio to gpu so slow?

### ISSUE THREAD

**Issue by trangtv57** [pytorch/audio#1607]
hi, tks for greate repo.
I have one question about how can i decrease time for load wav audio to gpu by torchaudio.  I want to load audio to gpu for take advantage of gpu for torchaudio transform feature by gpu:
First i have function to load audio with duration:

```python
def load_audio_with_duration_gpu(path_audio, start, duration):
	wf, sr_2 = torchaudio.load(path_audio)
        # transfer waveform to cuda if has gpu.
	# if torch.cuda.is_available():
	# 	wf = wf.cuda()
	audio_ft = wf.squeeze(0)
	return audio_ft[int(start * 16000):int((start + duration) * 16000)]
```

I try to benchmark with  loading 4000 audio (range 1-3s). If I don't transfer wf to cuda ( commend line code `if torch.cuda.is_available():`) it's take: 20s, and if i let waveform transfer to cuda, it's take 115s .
So what should i do for take advantage of using torchaudio compare with sox, or librosa , if i can't use gpu like this.
My understand can not correct, so can you correct me if i'm wrong. thank you.
additional info: my gpu 1080 Ti, and cuda: 10.1

**Comment by mthrok** [comment]
Hi @trangtv57 

Since you are loading audio and slicing, it is advantageous to give `frame_offset` and `num_frames`. That way, the function will stop decoding once enough frames are decoded. If you split loading and slicing, then the `load` function will load the whole audio.

Regarding CPU -> CUDA, loading audio and moving is the only way. If there is a library for decoding on GPU, then we can bind them (like `nvjpeg` for JPEG images), but AFAIK, there is no such library.

Having said that, to improved the overall throughput of your application, you can try parallelizing the loading process by creating dataset and use `Dataloader` with `num_process>1`. Then, by using `pin_memory=True`, you might get performance gain. https://pytorch.org/docs/stable/data.html

**Comment by trangtv57** [comment]
tks @mthrok , i just test phase inference for single sample pertime, so i don't using dataloader. I try to replace slicing audio using frame_offset and num_frames in torch_audio.load, but time is still don't decrease  much. anyway, tks you.
